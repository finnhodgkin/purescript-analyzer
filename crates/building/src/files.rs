use core::str;
use std::sync::Arc;

use id::Id;
use indexmap::IndexMap;
use rustc_hash::FxBuildHasher;

#[derive(Default)]
pub struct Files {
    files: IndexMap<Arc<str>, Arc<str>, FxBuildHasher>,
}

pub struct File;

pub type FileId = Id<File>;

impl Files {
    pub fn insert(&mut self, k: impl Into<Arc<str>>, v: impl Into<Arc<str>>) -> FileId {
        let k = k.into();
        let v = v.into();
        let (index, _) = self.files.insert_full(k, v);
        Id::from_raw(index)
    }

    pub fn id(&self, k: &str) -> Option<FileId> {
        self.files.get_full(k).map(|(index, _, _)| Id::from_raw(index))
    }

    pub fn path(&self, id: FileId) -> Arc<str> {
        let index = id.into();
        let (path, _) = self.files.get_index(index).expect("valid FileId");
        Arc::clone(path)
    }

    pub fn content(&self, id: FileId) -> Arc<str> {
        let index = id.into();
        let (_, contents) = self.files.get_index(index).expect("valid FileId");
        Arc::clone(contents)
    }
}

#[cfg(test)]
mod tests {
    use super::Files;

    #[test]
    fn test_basic() {
        let mut files = Files::default();

        let k = "src/Main.purs";
        let v = "module Main where\n\n";

        let id = files.insert(k, v);

        assert_eq!(files.id(k), Some(id));
        assert_eq!(files.path(id).as_ref(), k);
        assert_eq!(files.content(id).as_ref(), v);
    }
}
