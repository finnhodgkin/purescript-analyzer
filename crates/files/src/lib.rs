//! A virtual file system.
//!
//! This crate is adapted from rust-analyzer's `vfs` crate.
//!
//! In summary, this crate implements a virtual file system, [`Files`], used
//! to store the contents of the source files currently being analyzed. This
//! crate does not implement IO operations nor does it integrate directly
//! into the `salsa` database for the analyzer.
//!
//! Instead, the [`set_file_contents`] method handles file creation, deletion,
//! and modification inside the virtual file system. Likewise, changes are
//! logged and can be seen using the [`take_changes`] method.
//!
//! [`set_file_contents`]: Files::set_file_contents
//! [`take_changes`]: Files::take_changes

mod interner;

use std::{mem, path::PathBuf};

use interner::Interner;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FileId(u32);

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Change {
    Created,
    Deleted,
    Modified,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ChangedFile {
    pub file_id: FileId,
    pub change: Change,
}

#[derive(Default)]
pub struct Files {
    interner: Interner,
    files: Vec<Option<Vec<u8>>>,
    changes: Vec<ChangedFile>,
}

impl Files {
    /// Returns the [`FileID`] for a `path`.
    pub fn file_id(&self, path: PathBuf) -> Option<FileId> {
        self.interner.get(path).filter(|&file_id| self.get(file_id).is_some())
    }

    /// Returns the file path given a `file_id`.
    pub fn file_path(&self, file_id: FileId) -> PathBuf {
        self.interner.lookup(file_id).clone()
    }

    /// Returns the file contents given a `file_id`.
    pub fn file_contents(&self, file_id: FileId) -> &[u8] {
        self.get(file_id).as_deref().unwrap()
    }

    /// Sets the contents for a given `path`, with `None` meaning deletion.
    ///
    /// Returns `true` if the file was modified and saves the [change](ChangedFile).
    ///
    /// Allocates a new `FileId` if the path does not currently exist.
    pub fn set_file_contents(&mut self, path: PathBuf, mut contents: Option<Vec<u8>>) -> bool {
        let file_id = self.allocate_file_id(path);
        let change = match (self.get(file_id), &contents) {
            (None, None) => return false,
            (Some(q), Some(p)) if q == p => return false,
            (None, Some(_)) => Change::Created,
            (Some(_), None) => Change::Deleted,
            (Some(_), Some(_)) => Change::Modified,
        };
        if let Some(contents) = &mut contents {
            contents.shrink_to_fit();
        }
        *self.get_mut(file_id) = contents;
        self.changes.push(ChangedFile { file_id, change });
        true
    }

    /// Returns `true` if the file system logged changes.
    pub fn has_changes(&self) -> bool {
        !self.changes.is_empty()
    }

    /// Takes all changes logged so far by the file system.
    pub fn take_changes(&mut self) -> Vec<ChangedFile> {
        mem::take(&mut self.changes)
    }

    fn allocate_file_id(&mut self, path: PathBuf) -> FileId {
        let file_id = self.interner.intern(path);
        let idx = file_id.0 as usize;
        let len = self.files.len().max(idx + 1);
        self.files.resize_with(len, || None);
        file_id
    }

    fn get(&self, file_id: FileId) -> &Option<Vec<u8>> {
        &self.files[file_id.0 as usize]
    }

    fn get_mut(&mut self, file_id: FileId) -> &mut Option<Vec<u8>> {
        &mut self.files[file_id.0 as usize]
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use crate::{Change, ChangedFile, Files};

    #[test]
    fn api() {
        let mut files = Files::default();

        let path = PathBuf::from("./src/Main.purs");

        // Identity

        assert!(!files.set_file_contents(path.clone(), None));

        // Creation

        let contents = "module Main where".as_bytes();

        assert!(files.set_file_contents(path.clone(), Some(contents.to_vec())));
        let changes = files.take_changes();

        let file_id = files.file_id(path.clone()).unwrap();
        let file_contents = files.file_contents(file_id);

        assert_eq!(changes, &[ChangedFile { file_id, change: Change::Created }]);
        assert_eq!(contents, file_contents);

        // Identity
        assert!(!files.set_file_contents(path.clone(), Some(contents.to_vec())));

        // Modification

        let contents = "module Hello where".as_bytes();

        assert!(files.set_file_contents(path.clone(), Some(contents.to_vec())));
        let changes = files.take_changes();

        let file_contents = files.file_contents(file_id);

        assert_eq!(changes, &[ChangedFile { file_id, change: Change::Modified }]);
        assert_eq!(contents, file_contents);

        // Deletion

        assert!(files.set_file_contents(path.clone(), None));
        let changes = files.take_changes();
        assert_eq!(changes, &[ChangedFile { file_id, change: Change::Deleted }]);
    }
}
