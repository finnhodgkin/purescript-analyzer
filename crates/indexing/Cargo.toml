[package]
name = "indexing"
version = "0.1.0"
edition = "2021"

[dependencies]
fxhash = "0.2.1"
indexmap = "2.7.0"
la-arena = "0.3.1"
petgraph = "0.7.0"
rowan = "0.16.1"
rustc-hash = "2.1.0"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
