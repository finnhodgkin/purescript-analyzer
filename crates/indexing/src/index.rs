use std::sync::Arc;

use la_arena::{Arena, ArenaMap, Idx};
use rustc_hash::FxHashMap;
use smol_str::SmolStr;

use crate::source::*;

#[derive(Debug, PartialEq, Eq)]
pub enum TermItem {
    ClassMember { id: ClassMemberId },
    Constructor { id: DataConstructorId },
    Derive { id: DeriveId },
    Foreign { id: ForeignValueId },
    Instance { id: InstanceId },
    Operator { id: InfixId },
    Value { signature: Option<ValueSignatureId>, equations: Vec<ValueEquationId> },
}

pub type TermItemId = Idx<TermItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum TypeItem {
    Data {
        signature: Option<DataSignatureId>,
        equation: Option<DataEquationId>,
        role: Option<TypeRoleId>,
    },
    Newtype {
        signature: Option<NewtypeSignatureId>,
        equation: Option<NewtypeEquationId>,
        role: Option<TypeRoleId>,
    },
    Synonym {
        signature: Option<TypeSignatureId>,
        equation: Option<TypeEquationId>,
    },
    Class {
        signature: Option<ClassSignatureId>,
        declaration: Option<ClassDeclarationId>,
    },
    Foreign {
        id: ForeignDataId,
    },
    Operator {
        id: InfixId,
    },
}

pub type TypeItemId = Idx<TypeItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum ImplicitItems {
    Everything,
    Enumerated(Arc<[SmolStr]>),
}

pub type ImportedTerms = FxHashMap<SmolStr, ImportItemId>;
pub type ImportedTypes = FxHashMap<SmolStr, (ImportItemId, Option<ImplicitItems>)>;

#[derive(Debug, PartialEq, Eq)]
pub struct ImportedItems {
    pub terms: ImportedTerms,
    pub types: ImportedTypes,
    pub exported: bool,
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Index {
    term_item: Arena<TermItem>,
    type_item: Arena<TypeItem>,
    term_export: ArenaMap<TermItemId, ExportItemId>,
    type_export: ArenaMap<TypeItemId, ExportItemId>,
    alias_nominal: FxHashMap<SmolStr, Vec<ImportId>>,
    imported_items: FxHashMap<ImportId, ImportedItems>,
    term_nominal: FxHashMap<SmolStr, TermItemId>,
    type_nominal: FxHashMap<SmolStr, TypeItemId>,
}

impl std::ops::Index<TermItemId> for Index {
    type Output = TermItem;

    fn index(&self, id: TermItemId) -> &TermItem {
        &self.term_item[id]
    }
}

impl std::ops::Index<TypeItemId> for Index {
    type Output = TypeItem;

    fn index(&self, id: TypeItemId) -> &TypeItem {
        &self.type_item[id]
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Relational {
    data: Vec<(TypeItemId, TermItemId)>,
    class: Vec<(TypeItemId, TermItemId)>,
    chain: Vec<(InstanceChainId, InstanceId)>,
    instance: Vec<(InstanceId, InstanceMemberId)>,
}

impl Index {
    pub(crate) fn insert_term_item(&mut self, k: Option<SmolStr>, v: TermItem) -> TermItemId {
        let id = self.term_item.alloc(v);
        let Some(k) = k else { return id };
        self.term_nominal.insert(k, id);
        id
    }

    pub(crate) fn term_item_mut(&mut self, k: &str) -> Option<(&mut TermItem, TermItemId)> {
        let &id = self.term_nominal.get(k)?;
        Some((&mut self.term_item[id], id))
    }

    pub(crate) fn export_term_item(&mut self, item_id: TermItemId, id: ExportItemId) {
        self.term_export.insert(item_id, id);
    }

    pub(crate) fn term_item_export(&self, id: TermItemId) -> Option<ExportItemId> {
        self.term_export.get(id).copied()
    }

    pub(crate) fn insert_type_item(&mut self, k: Option<SmolStr>, v: TypeItem) -> TypeItemId {
        let id = self.type_item.alloc(v);
        let Some(k) = k else { return id };
        self.type_nominal.insert(k, id);
        id
    }

    pub(crate) fn type_item_mut(&mut self, k: &str) -> Option<(&mut TypeItem, TypeItemId)> {
        let &id = self.type_nominal.get(k)?;
        Some((&mut self.type_item[id], id))
    }

    pub(crate) fn export_type_item(&mut self, item_id: TypeItemId, id: ExportItemId) {
        self.type_export.insert(item_id, id);
    }

    pub(crate) fn type_item_export(&self, id: TypeItemId) -> Option<ExportItemId> {
        self.type_export.get(id).copied()
    }

    pub(crate) fn insert_import_alias(&mut self, k: SmolStr, v: ImportId) {
        self.alias_nominal.entry(k).or_default().push(v);
    }

    pub(crate) fn export_import_alias(&mut self, k: &str) {
        if let Some(imports) = self.alias_nominal.get(k) {
            for import in imports {
                if let Some(items) = self.imported_items.get_mut(import) {
                    items.exported = true;
                }
            }
        }
    }

    pub(crate) fn insert_imported_items(&mut self, k: ImportId, v: ImportedItems) {
        self.imported_items.insert(k, v);
    }

    pub fn lookup_import_alias(&self, k: &str) -> Option<&[ImportId]> {
        Some(self.alias_nominal.get(k)?)
    }

    pub fn iter_term_item(&self) -> impl Iterator<Item = (TermItemId, &TermItem)> {
        self.term_item.iter()
    }

    pub fn iter_type_item(&self) -> impl Iterator<Item = (TypeItemId, &TypeItem)> {
        self.type_item.iter()
    }
}

impl Relational {
    pub(crate) fn insert_data_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.data.push((type_id, term_id));
    }

    pub(crate) fn insert_class_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.class.push((type_id, term_id));
    }

    pub(crate) fn insert_chain_relation(&mut self, c_id: InstanceChainId, i_id: InstanceId) {
        self.chain.push((c_id, i_id));
    }

    pub(crate) fn insert_instance_relation(&mut self, i_id: InstanceId, m_id: InstanceMemberId) {
        self.instance.push((i_id, m_id));
    }

    pub fn constructors_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.data.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }

    pub fn class_members_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.class.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum ItemKind {
    DataSignature(DataSignatureId),
    DataEquation(DataEquationId),

    NewtypeSignature(NewtypeSignatureId),
    NewtypeEquation(NewtypeEquationId),

    Constructor(DataConstructorId),

    SynonymSignature(TypeSignatureId),
    SynonymEquation(TypeEquationId),

    ClassSignature(ClassSignatureId),
    ClassDeclaration(ClassDeclarationId),
    ClassMember(ClassMemberId),

    ValueSignature(ValueSignatureId),
    ValueEquation(ValueEquationId),

    ForeignData(ForeignDataId),
    ForeignValue(ForeignValueId),
    Operator(InfixId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum ExistingKind {
    Term(TermItemId),
    Type(TypeItemId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum IndexError {
    DuplicateItem { kind: ItemKind, existing: ExistingKind },
    MismatchedItem { kind: ItemKind, existing: ExistingKind },
    InvalidRole { id: TypeRoleId, existing: Option<TypeItemId> },
    InvalidExport { id: ExportItemId },
    DuplicateExport { id: ExportItemId, existing: ExportItemId },
    DuplicateImport { id: ImportItemId, existing: ImportItemId },
}
