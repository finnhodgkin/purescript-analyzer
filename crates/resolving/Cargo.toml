[package]
name = "resolving"
version = "0.1.0"
edition = "2024"

[dependencies]
files = { version = "0.1.0", path = "../files" }
indexing = { version = "0.1.0", path = "../indexing" }
la-arena = "0.3.1"
rustc-hash = "2.1.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
rowan = "0.16.1"
