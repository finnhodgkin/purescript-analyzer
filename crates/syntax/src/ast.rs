#[macro_use]
mod macros;

mod binder;
mod binding;
mod common;
mod declaration;
mod module;
mod name;

pub use binder::*;
pub use binding::*;
pub use common::*;
pub use declaration::*;
pub use module::*;
pub use name::*;
