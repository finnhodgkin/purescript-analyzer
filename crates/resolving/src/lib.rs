use std::{collections::hash_map::Entry, sync::Arc};

use files::FileId;
use indexing::{FullModuleIndex, Index, TermItemId, TypeItemId};
use rustc_hash::FxHashMap;
use smol_str::SmolStr;

pub trait External {
    fn index(&mut self, id: FileId) -> Arc<FullModuleIndex>;
}

pub fn exports(external: &mut impl External, file: FileId) {
    let mut state = State::default();
    let index = &external.index(file).index;
    match index.export_kind {
        indexing::ExportKind::Implicit => {
            implicit_module_exports(&mut state, file, index);
        }
        indexing::ExportKind::Explicit => {
            explicit_module_exports(&mut state, file, index, false);
        }
        indexing::ExportKind::ExplicitSelf => {
            explicit_module_exports(&mut state, file, index, true);
        }
    }
    dbg!(&state);
}

#[derive(Debug, PartialEq, Eq)]
pub enum Error {
    ExistingTerm { existing: (FileId, TermItemId), duplicate: (FileId, TermItemId) },
    ExistingType { existing: (FileId, TypeItemId), duplicate: (FileId, TypeItemId) },
}

#[derive(Debug, Default)]
struct State {
    errors: Vec<Error>,
    terms: FxHashMap<SmolStr, (FileId, TermItemId)>,
    types: FxHashMap<SmolStr, (FileId, TypeItemId)>,
}

impl State {
    fn add_term(&mut self, name: &str, file: FileId, id: TermItemId) {
        let k = SmolStr::from(name);
        match self.terms.entry(k) {
            Entry::Occupied(o) => {
                let existing = *o.get();
                let duplicate = (file, id);
                self.errors.push(Error::ExistingTerm { existing, duplicate });
            }
            Entry::Vacant(v) => {
                v.insert((file, id));
            }
        }
    }

    fn add_type(&mut self, name: &str, file: FileId, id: TypeItemId) {
        let k = SmolStr::from(name);
        match self.types.entry(k) {
            Entry::Occupied(o) => {
                let existing = *o.get();
                let duplicate = (file, id);
                self.errors.push(Error::ExistingType { existing, duplicate });
            }
            Entry::Vacant(v) => {
                v.insert((file, id));
            }
        }
    }
}

fn implicit_module_exports(state: &mut State, file: FileId, index: &Index) {
    for (name, id) in index.iter_term_nominal() {
        state.add_term(name, file, id);
    }
    for (name, id) in index.iter_type_nominal() {
        state.add_type(name, file, id);
    }
}

fn explicit_module_exports(state: &mut State, file: FileId, index: &Index, including_self: bool) {
    if including_self {
        implicit_module_exports(state, file, index);
    }
}
