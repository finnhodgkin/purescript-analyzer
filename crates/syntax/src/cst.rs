use rowan::ast::{AstChildren, AstNode};

use crate::{SyntaxKind, SyntaxToken};

#[macro_use]
mod macros;

create_cst_struct!(Module, ModuleHeader, ExportList);

create_cst_enum!(
    ExportItem
        | ExportValue
        | ExportClass
        | ExportType
        | ExportOperator
        | ExportTypeOperator
        | ExportModule
);

create_cst_struct!(ModuleImports, ImportStatement, ImportList, ImportAlias);

create_cst_enum!(
    ImportItem // Comment to fix formatting :P
        | ImportValue
        | ImportClass
        | ImportType
        | ImportOperator
        | ImportTypeOperator
);

create_cst_enum!(TypeItems | TypeItemsAll | TypeItemsList);

create_cst_struct!(ModuleStatements);

create_cst_enum!(
    Declaration
        | ValueSignature
        | ValueEquation
        | InfixDeclaration
        | TypeRoleDeclaration
        | InstanceChain
        | InstanceDeclaration
        | InstanceSignatureStatement
        | InstanceEquationStatement
        | TypeSynonymSignature
        | TypeSynonymEquation
        | ClassSignature
        | ClassDeclaration
        | ClassMemberStatement
        | ForeignImportDataDeclaration
        | ForeignImportValueDeclaration
        | NewtypeSignature
        | NewtypeEquation
        | DataSignature
        | DataEquation
        | DeriveDeclaration
);

associated_declarations!(
    ClassDeclaration where ClassMemberStatement
);

associated_declarations!(
    InstanceDeclaration where InstanceSignatureStatement | InstanceEquationStatement
);

create_cst_struct!(
    TypeRole,
    ClassConstraints,
    ClassHead,
    ClassFunctionalDependencies,
    ClassStatements
);

create_cst_enum!(
    FunctionalDependency | FunctionalDependencyDetermined | FunctionalDependencyDetermines
);

create_cst_struct!(InstanceName, InstanceConstraints, InstanceHead, InstanceStatements);

create_cst_struct!(DataConstructor);

create_cst_enum!(
    Type // Comment to fix formatting :P
        | TypeApplicationChain
        | TypeArrow
        | TypeConstrained
        | TypeConstructor
        | TypeForall
        | TypeHole
        | TypeInteger
        | TypeKinded
        | TypeOperator
        | TypeOperatorChain
        | TypeString
        | TypeVariable
        | TypeVariableBinding
        | TypeWildcard
        | TypeRecord
        | TypeRow
        | TypeParenthesized
);

create_cst_enum!(
    Binder
        | BinderTyped
        | BinderOperatorChain
        | BinderInteger
        | BinderNumber
        | BinderConstructor
        | BinderVariable
        | BinderNamed
        | BinderWildcard
        | BinderString
        | BinderChar
        | BinderTrue
        | BinderFalse
        | BinderArray
        | BinderRecord
        | BinderParenthesized
);

create_cst_struct!(
    FunctionBinders,
    Unconditional,
    WhereExpression,
    Conditionals,
    PatternGuarded,
    PatternGuards
);

create_cst_enum!(PatternGuard | PatternGuardBinder | PatternGuardExpression);

create_cst_enum!(
    Expression
        | ExpressionTyped
        | ExpressionOperatorChain
        | ExpressionInfixChain
        | ExpressionTick
        | ExpressionNegate
        | ExpressionApplicationChain
        | ExpressionTypeArgument
        | ExpressionTermArgument
        | ExpressionIfThenElse
        | ExpressionLetIn
        | ExpressionLambda
        | ExpressionCaseOf
        | ExpressionDo
        | ExpressionAdo
        | ExpressionConstructor
        | ExpressionVariable
        | ExpressionOperatorName
        | ExpressionSection
        | ExpressionHole
        | ExpressionString
        | ExpressionChar
        | ExpressionTrue
        | ExpressionFalse
        | ExpressionInteger
        | ExpressionNumber
        | ExpressionArray
        | ExpressionRecord
        | ExpressionParenthesized
        | ExpressionRecordAccess
        | ExpressionRecordUpdate
);

create_cst_struct!(CaseTrunk, CaseBranches, CaseBranchBinders, CaseBranch);

create_cst_enum!(DoStatement | DoStatementBind | DoStatementLet | DoStatementDiscard);

create_cst_enum!(LetBinding | LetBindingPattern | LetBindingSignature | LetBindingEquation);

create_cst_enum!(RecordItem | RecordField | RecordPun);

create_cst_struct!(RecordUpdates);

create_cst_enum!(RecordUpdate | RecordUpdateLeaf | RecordUpdateBranch);

impl Module {
    pub fn statements(&self) -> Option<ModuleStatements> {
        self.syntax().children().find_map(ModuleStatements::cast)
    }
}

impl ModuleStatements {
    pub fn children(&self) -> impl Iterator<Item = Declaration> {
        self.syntax().children().filter_map(Declaration::cast)
    }
}

impl InstanceChain {
    pub fn instance_declarations(&self) -> AstChildren<InstanceDeclaration> {
        rowan::ast::support::children(self.syntax())
    }
}

impl InstanceDeclaration {
    pub fn instance_head(&self) -> Option<InstanceHead> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_name(&self) -> Option<InstanceName> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_statements(&self) -> Option<InstanceStatements> {
        rowan::ast::support::child(self.syntax())
    }
}

impl InstanceName {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl InstanceHead {
    pub fn type_name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::UPPER)
    }
}

impl InstanceStatements {
    pub fn children(&self) -> AstChildren<InstanceDeclarationStatement> {
        rowan::ast::support::children(self.syntax())
    }
}

impl ValueSignature {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl ValueEquation {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}
