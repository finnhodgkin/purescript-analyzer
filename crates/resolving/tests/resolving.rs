use std::sync::Arc;

use files::FileId;
use indexing::FullModuleIndex;
use la_arena::RawIdx;
use lowering::FullModuleLower;
use resolving::{External, FullResolvedModule};
use rowan::ast::AstNode;
use syntax::cst;

pub const MAIN: &str = include_str!("fixtures/Main.txt");
pub const LIB: &str = include_str!("fixtures/Lib.txt");
pub const INTERNAL: &str = include_str!("fixtures/Internal.txt");

macro_rules! file {
    ($id:expr) => {
        FileId::from_raw(RawIdx::from_u32($id))
    };
}

pub fn parse(id: FileId) -> cst::Module {
    let id = id.into_raw().into_u32();
    let source = match id {
        0 => MAIN,
        1 => LIB,
        2 => INTERNAL,
        _ => unreachable!(),
    };

    let lexed = lexing::lex(source);
    let tokens = lexing::layout(&lexed);

    let (module, _) = parsing::parse(&lexed, &tokens);

    cst::Module::cast(module).unwrap()
}

struct TestExternal;

impl External for TestExternal {
    fn indexed(&mut self, id: FileId) -> Arc<FullModuleIndex> {
        let module = parse(id);
        let index = indexing::index_module(&module);
        Arc::new(index)
    }

    fn lowered(&mut self, id: FileId) -> Arc<FullModuleLower> {
        let module = parse(id);
        let index = self.indexed(id);
        let lowered =
            lowering::lower_module(&module, &index.index, &index.relational, &index.source);
        Arc::new(lowered)
    }

    fn resolved(&mut self, id: FileId) -> Arc<FullResolvedModule> {
        let exports = resolving::resolve_module(self, id);
        Arc::new(exports)
    }

    fn file_id(&mut self, name: &str) -> FileId {
        match name {
            "Main" => file!(0),
            "Lib" => file!(1),
            "Internal" => file!(2),
            _ => unreachable!(),
        }
    }
}

#[test]
fn test_basic() {
    let mut external = TestExternal;

    let lowered = external.lowered(file!(0));
    let resolved = external.resolved(file!(0));

    for (id, deferred) in lowered.graph.deferred() {
        match deferred.domain {
            lowering::ResolutionDomain::Term => {
                if let Some(name) = &deferred.name {
                    let r = resolved.lookup_term(name);
                    println!("{:?} = {:?}", id, r);
                }
            }
            lowering::ResolutionDomain::Type => {
                if let Some(name) = &deferred.name {
                    let r = resolved.lookup_type(name);
                    println!("{:?} = {:?}", id, r);
                }
            }
        }
    }
}
