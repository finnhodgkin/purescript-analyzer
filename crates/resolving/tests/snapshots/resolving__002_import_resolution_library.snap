---
source: crates/resolving/tests/resolving.rs
expression: report
snapshot_kind: text
---
module Library

Unqualified Imports:

Qualified Imports:

Exported Terms:
  - libFn
  - hideMe

Exported Types:

Local Terms:
  - libFn
  - hideMe

Local Types:
